### PR TITLE
Size compression

### PR DESCRIPTION
**Pull request by Fatalll**

### CODE REVIEW

File: buildSrc/src/main/kotlin/tanvd/grazi/Dependencies.kt [hunk 1]
```diff
@@ -20,10 +20,23 @@ fun ExternalModuleDependency.ltExcludes() {
     exclude("javax.xml.bind", "jaxb-api")
     exclude("org.glassfish.jaxb", "jaxb-runtime")
     exclude("com.fasterxml.jackson.core", "jackson-databind")
+    exclude("com.esotericsoftware.kryo", "kryo")
 
-    // used only in languagetool-core test
+    exclude("org.apache.commons", "commons-lang3")
+//    exclude("org.apache.commons", "commons-csv")
```
Comment (TanVD) [inline buildSrc/src/main/kotlin/tanvd/grazi/Dependencies.kt:26]: So, what's with commented parts?

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package tanvd.grazi
+
+import com.intellij.openapi.application.ApplicationManager
+import org.jetbrains.org.objectweb.asm.*
+import org.jetbrains.org.objectweb.asm.Opcodes.ASM5
+
+
+class EnglishChonker(cv: ClassVisitor) : ClassVisitor(ASM5, cv) {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Let's rename it to `EnglishChunkerInstrumentation` and move into LangToolInstrumentation object

File: plugin/src/main/kotlin/tanvd/grazi/spellcheck/GraziSpellchecker.kt [hunk 3]
```diff
@@ -77,7 +85,12 @@ object GraziSpellchecker : GraziStateLifecycle {
      * Note, that casing and plural typos are ignored.
      */
     private fun check(word: String, project: Project, language: Language) = buildSet<Typo> {
-        val typo = tryRun { checker.check(word) }?.firstOrNull()?.let { Typo(it, checkerLang, 0) }
+        val typo = try {
+            checker.check(word)
+        } catch (e: Throwable) {
+            logger.trace("GraziSpellchecker exception", e)
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/spellcheck/GraziSpellchecker.kt]: `Exception occurred during check in GraziSpellchecker` ;-)

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    override fun visitMethod(access: Int, name: String?, descriptor: String?, signature: String?, exceptions: Array<out String>?): MethodVisitor {
+        val mv = cv.visitMethod(access, name, descriptor, signature, exceptions)
+        return if (name.equals("getChunker")) { GetChunkerMethodVisitor(mv) } else { mv }
+    }
+
+    class GetChunkerMethodVisitor(private val visitor: MethodVisitor) : MethodVisitor(ASM5, null) {
+        override fun visitCode() {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: `=with(visitor) {...}`

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                } else {
+                    invoke(GraziPlugin.classLoader, "org.languagetool.language.English", cls, 0, cls.size)
+                }
+            }
+        }
+    }
+
+    override fun visitMethod(access: Int, name: String?, descriptor: String?, signature: String?, exceptions: Array<out String>?): MethodVisitor {
+        val mv = cv.visitMethod(access, name, descriptor, signature, exceptions)
+        return if (name.equals("getChunker")) { GetChunkerMethodVisitor(mv) } else { mv }
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: `name == "getChunker"`?

Comment (TanVD) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: And unneeded braces. Also you can use `ifTrue`

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            val reader = ClassReader(stream)
+            val writer = ClassWriter(ClassWriter.COMPUTE_MAXS)
+            val visitor = EnglishChonker(writer)
+            reader.accept(visitor, 0)
+
+            val cls = writer.toByteArray()
+            with(ClassLoader::class.java.getDeclaredMethod("defineClass", String::class.java, ByteArray::class.java, Int::class.java, Int::class.java)) {
+                isAccessible = true
+
+                if (ApplicationManager.getApplication().isUnitTestMode) {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: It looks like we should handle this on side of `GraziPlugin.classLoader` :)

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.jetbrains.org.objectweb.asm.Opcodes.ASM5
+
+
+class EnglishChonker(cv: ClassVisitor) : ClassVisitor(ASM5, cv) {
+    companion object {
+        fun reloadEnglish() {
+            val stream = GraziPlugin::class.java.getResourceAsStream("/org/languagetool/language/English.class")
+            val reader = ClassReader(stream)
+            val writer = ClassWriter(ClassWriter.COMPUTE_MAXS)
+            val visitor = EnglishChonker(writer)
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: inline to call site

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import com.intellij.openapi.application.ApplicationManager
+import org.jetbrains.org.objectweb.asm.*
+import org.jetbrains.org.objectweb.asm.Opcodes.ASM5
+
+
+class EnglishChonker(cv: ClassVisitor) : ClassVisitor(ASM5, cv) {
+    companion object {
+        fun reloadEnglish() {
+            val stream = GraziPlugin::class.java.getResourceAsStream("/org/languagetool/language/English.class")
+            val reader = ClassReader(stream)
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Inline to call site

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.intellij.openapi.application.ApplicationManager
+import org.jetbrains.org.objectweb.asm.*
+import org.jetbrains.org.objectweb.asm.Opcodes.ASM5
+
+
+class EnglishChonker(cv: ClassVisitor) : ClassVisitor(ASM5, cv) {
+    companion object {
+        fun reloadEnglish() {
+            val stream = GraziPlugin::class.java.getResourceAsStream("/org/languagetool/language/English.class")
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Hm, should not we do it throught some interface , like `Resources.stream(filePath)`? To fix classloader for resources resolution

**Review by TanVD — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/tanvd/grazi/Dependencies.kt (hunk 1 above)
Comment (Fatalll) [reply buildSrc/src/main/kotlin/tanvd/grazi/Dependencies.kt:26]: They are not in IDEA, but somehow it works without them...

Comment (Fatalll) [reply buildSrc/src/main/kotlin/tanvd/grazi/Dependencies.kt:26]: Gonna inspect it in the next pull request, where we removing LangDetector and html

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 2 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/spellcheck/GraziSpellchecker.kt (hunk 3 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/spellcheck/GraziSpellchecker.kt]: Fixed

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 4 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 5 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 6 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 7 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 8 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt (hunk 9 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/EnglishChonker.kt]: Create our Resources object

File: plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt [hunk 10]
```diff
@@ -69,4 +80,35 @@ object LangToolInstrumentation {
...
         }
     }
+
+    fun reloadEnglish() {
+        val reader = ClassReader(Resources.getStream("/org/languagetool/language/English.class"))
+        val writer = ClassWriter(ClassWriter.COMPUTE_MAXS)
+        reader.accept(EnglishChunkerInstrumentation(writer), 0)
+
+        val cls = writer.toByteArray()
+        with(ClassLoader::class.java.getDeclaredMethod("defineClass", String::class.java, ByteArray::class.java, Int::class.java, Int::class.java)) {
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt]: Huh, OK, found it here `https://asm.ow2.io/faq.html#Q5` :)

Comment (TanVD) [reply plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt]: Can we make this and others parts where we get methods via reflection extension functions and fields? For example define `ClassLoader.defineClass(...) = this code`

File: plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt [hunk 11]
```diff
@@ -69,4 +80,35 @@ object LangToolInstrumentation {
...
+            isAccessible = true
+            invoke(GraziPlugin.classLoader, "org.languagetool.language.English", cls, 0, cls.size)
+        }
+    }
+
+    private class EnglishChunkerInstrumentation(cv: ClassVisitor) : ClassVisitor(Opcodes.ASM5, cv) {
+        override fun visitMethod(access: Int, name: String?, descriptor: String?, signature: String?, exceptions: Array<out String>?): MethodVisitor {
+            val mv = cv.visitMethod(access, name, descriptor, signature, exceptions)
```
Comment (TanVD) [inline plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt]: `cv - classVisitor` and `mv - methodVisitor` -- let's be a little bit less c-plus-plus programmers)

**Review by TanVD — CHANGES_REQUESTED** [review]

File: plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt (hunk 10 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt]: Done

File: plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt (hunk 11 above)
Comment (Fatalll) [reply plugin/src/main/kotlin/tanvd/grazi/utils/JLangUtils.kt]: cv is protected field, but ok :)

**Pull request merged**